blocks/search.py

```diff
@@ -0,0 +1,401 @@
+                           for value in outputs]
+            return OrderedDict(zip(names, outputs))
+        return wrapper
+    if len(args) == 1 and callable(args[0]):
+        # No arguments, this is the decorator
+        # Set default values for the arguments
+        add_time_dim = True
+        return _trace(args[0])
+    else:
+        # This is just returning the decorator
+        add_time_dim = kwargs['add_time_dim']
+        return _trace
+
+
+class BeamSearch(Search):
+    """Beam search.
+
+    Parameters
+    ----------
+    beam_size : int
+        Size of beam.
+    batch_size : int
+        Size of input batch.
+    sequence_generator : sequence generator
+        Sequence generator brick.
+    attended : theano variable
+        Theano variable for attended in sequence generator.
+    attended_mask : theano variable
+        Theano variable for attended mask in sequence generator.
+    inputs_dict : dict
+        Dictionary of inputs {name: theano variable for input}. The
+        functions will be constructed with these inputs.
+
+    """
+    def __init__(self, beam_size, batch_size, sequence_generator, attended,
+                 attended_mask, inputs_dict):
+        super(BeamSearch, self).__init__(sequence_generator)
+        self.beam_size = beam_size
+        self.batch_size = batch_size
+        self.sequence_generator = sequence_generator
+        self.attended = attended
+        self.attended_mask = attended_mask
+        self.inputs_dict = inputs_dict
+        self.generate_names = sequence_generator.generate_outputs()
+        self.init_computer = None
+        self.next_computer = None
+        self.attended_computer = None
+        self.initial_state_computer = None
+        self.real_batch_size = batch_size * beam_size
+        self.state_names = (sequence_generator.state_names +
+                            sequence_generator.glimpse_names +
+                            ['outputs'])
+
+    def compile(self, *args, **kwargs):
+        """Compiles functions for beam search."""
+        generator = self.sequence_generator
+        attended = self.attended
+        attended_mask = self.attended_mask
+
+        self.attended_computer = function(self.inputs_dict.values(),
+                                          [attended, attended_mask],
+                                          on_unused_input='ignore')
+
+        initial_states = OrderedDict()
+        for name in self.state_names:
+            initial_states[name] = generator.initial_state(
+                name,
+                self.real_batch_size,
+                attended=attended)
+
+        self.initial_state_computer = function([attended, attended_mask],
+                                               initial_states.values(),
+                                               on_unused_input='ignore')
+        # Construct initial values
+        init_generated = generator.generate(attended=attended,
```

Comment:
I am sorry I still do not understand why you need separate `init_generated` and `next_generated`.